src/sources/mssql/mssql-schema.lisp

```diff
@@ -61,46 +39,55 @@
   "Associate internal table type symbol with what's found in MS SQL
   information_schema.tables.table_type column.")
 
-(defun filter-list-to-where-clause (filter-list
-                                    &optional
-                                      not
-                                      (schema-col "table_schema")
-                                      (table-col  "table_name"))
+(defmethod filter-list-to-where-clause ((mssql copy-mssql)
+                                        filter-list
+                                        &key
+                                          not
+                                          (schema-col "table_schema")
+                                          (table-col  "table_name"))
   "Given an INCLUDING or EXCLUDING clause, turn it into a MS SQL WHERE clause."
   (loop :for (schema . table-name-list) :in filter-list
      :append (mapcar (lambda (table-name)
                        (format nil "(~a = '~a' and ~a ~:[~;NOT ~]LIKE '~a')"
                                schema-col schema table-col not table-name))
                      table-name-list)))
 
-(defun list-all-columns (catalog
-                         &key
-			   (table-type :table)
-                           including
-                           excluding
-			 &aux
-			   (table-type-name (cdr (assoc table-type *table-type*))))
+(defmethod fetch-columns ((catalog catalog)
+                          (mssql copy-mssql)
+                          &key
+                            (table-type :table)
+                            including
+                            excluding
+                          &aux
+                            (table-type-name
+                             (cdr (assoc table-type *table-type*))))
   (loop
      :for (schema-name table-name name type default nullable identity
                        character-maximum-length
                        numeric-precision numeric-precision-radix numeric-scale
                        datetime-precision
                        character-set-name collation-name)
      :in
-     (mssql-query (format nil
-                          (sql "/mssql/list-all-columns.sql")
-                          (db-name *mssql-db*)
-                          table-type-name
-                          including     ; do we print the clause?
-                          (filter-list-to-where-clause including
-                                                       nil
-                                                       "c.table_schema"
-                                                       "c.table_name")
-                          excluding     ; do we print the clause?
-                          (filter-list-to-where-clause excluding
-                                                       t
-                                                       "c.table_schema"
-                                                       "c.table_name")))
+     (let ((including
```

Comment:
`LET` inside `LOOP` is not idiomatic - local variables should be declared via the `:WITH` clause if they do not depend on the results of the iteration.

My suggestion that should be equivalent (but let's test it anyway):

```lisp
(defmethod fetch-columns ((catalog catalog)
                          (mssql copy-mssql)
                          &key
                            (table-type :table)
                            including
                            excluding
                          &aux
                            (table-type-name
                             (cdr (assoc table-type *table-type*))))
  (loop
    :with including := (filter-list-to-where-clause mssql
                                                    including
                                                    :not nil
                                                    :schema-col "c.table_schema"
                                                    :table-col "c.table_name")
    :with excluding := (filter-list-to-where-clause mssql
                                                    excluding
                                                    :not t
                                                    :schema-col "c.table_schema"
                                                    :table-col "c.table_name")
    :for (schema-name table-name name type default nullable identity
                      character-maximum-length
                      numeric-precision numeric-precision-radix numeric-scale
                      datetime-precision
                      character-set-name collation-name)
      :in (mssql-query (format nil
                               (sql "/mssql/list-all-columns.sql")
                               (db-name *mssql-db*)
                               table-type-name
                               including   ; do we print the clause?
                               including
                               excluding   ; do we print the clause?
                               excluding))
    :do
       (let* ((schema     (maybe-add-schema catalog schema-name))
              (table      (maybe-add-table schema table-name))
              (field
                (make-mssql-column
                 schema-name table-name name type default nullable
                 (eq 1 identity)
                 character-maximum-length
                 numeric-precision numeric-precision-radix numeric-scale
                 datetime-precision
                 character-set-name collation-name)))
         (add-field table field))
    :finally (return catalog)))
```

Reply:
Let's try and be more idiomatic then! Thanks for the example, I've done it that way now. Only change is that since I've been beaten in the past with binding scopes and overloads in LOOP, I changed the bindings we use at the :with level. The including/excluding clauses from the pgloader command grammar are turned into SQL where clause parts, now named incl-where and excl-where.